rbac/management/role/v2_model.py

```diff
@@ -0,0 +1,193 @@
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+"""Model for role V2 management."""
+
+import uuid_utils.compat as uuid
+from django.db import models
+from django.utils import timezone
+from management.models import Permission, Role
+from management.rbac_fields import AutoDateTimeField
+from rest_framework import serializers
+
+from api.models import TenantAwareModel
+
+
+class RoleV2(TenantAwareModel):
+    """V2 Role model."""
+
+    class Types(models.TextChoices):
+        CUSTOM = "custom"
+        SEEDED = "seeded"
+        PLATFORM = "platform"
+
+    uuid = models.UUIDField(default=uuid.uuid7, editable=False, unique=True, null=False)
+    name = models.CharField(max_length=150, null=False, blank=False)
+    display_name = models.CharField(max_length=150, default="")
+    description = models.TextField(null=True)
+    type = models.CharField(
+        choices=Types.choices, max_length=20, null=False, blank=False, db_index=True, default=Types.CUSTOM
+    )
+    permissions = models.ManyToManyField(Permission, related_name="v2_roles")
+    children = models.ManyToManyField("self", related_name="parents", symmetrical=False)
+    v1_source = models.ForeignKey(Role, null=True, related_name="v2_roles", on_delete=models.SET_NULL)
+    version = models.PositiveIntegerField(default=1)
+    created = models.DateTimeField(default=timezone.now)
+    modified = AutoDateTimeField(default=timezone.now)
+
+    def save(self, *args, **kwargs):
+        """Ensure that display_name is populated on save."""
+        if not self.display_name:
+            self.display_name = self.name
+        super().save(*args, **kwargs)
+
+    class Meta:
+        ordering = ["name", "modified"]
+        constraints = [
+            models.UniqueConstraint(fields=["name", "tenant"], name="unique role v2 name per tenant"),
+            models.UniqueConstraint(fields=["display_name", "tenant"], name="unique role v2 display name per tenant"),
+        ]
+
+
+class TypedRoleV2Manager(models.Manager):
+    """Manager for RoleV2 with a specific type."""
+
+    _role_type: str
+
+    def __init__(self, role_type: str):
+        """Initialize the manager."""
+        super().__init__()
+        self._role_type = role_type
+
+    def get_queryset(self):
+        """Get the queryset for the specific type."""
+        return super().get_queryset().filter(type=self._role_type)
+
+
+class CustomRoleV2(RoleV2):
+    """V2 Custom role model."""
+
+    class Meta:
+        proxy = True
+
+    _expected_type = RoleV2.Types.CUSTOM
+    objects = TypedRoleV2Manager(role_type=_expected_type)
+
+    def __init__(self, *args, **kwargs):
```

Comment:
we can have another class called TypeValidatedRoleV2Mixin and put these __init__ and save method into it, so we don't need them in the subclasses

Reply:
yes that was something I was considering and then I decided this way because in the future the methods can differ for different types ... but I dont have strong preference here